### PR TITLE
Added feature flag for asynchronous export

### PR DESCRIPTION
**Pull request by DebajitDas**
Fixes # Adds feature flag for asynchronous export.
Changes for max export and wait on async export not triggered by shutdown/force-flush are pending 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by lalitb** [comment]
LGTM in general,  @owent my want to have a look into this, specifically the batch span/log processor part. Also, we should add a CI action to test the async feature ( e.g, add another test similar to [bazel_test](https://github.com/open-telemetry/opentelemetry-cpp/blob/a7e814a632fdaee0c76246358ffad811ac06a7e9/.github/workflows/ci.yml#L131) which builds using ENABLE_ASYNC_EXPORT macro).

File: CMakeLists.txt
Comment (owent) [inline CMakeLists.txt:231]: We should also add INTERFACE definitions into [api/CMakeLists.txt](api/CMakeLists.txt).

```cmake
if(WITH_LOGS_PREVIEW)
  target_compile_definitions(opentelemetry_api INTERFACE ENABLE_ASYNC_EXPORT)
endif()
```

So that when we use optentelemetry-cpp as a cmake module(`find_package(opentelemetry-cpp CONFIG)`) , this definition can be auto added into all targets that direct or indirectly depend on `opentelemetry-cpp`.

I find there are severval missing definitions in [api/CMakeLists.txt](api/CMakeLists.txt) and be add by `add_definitions()`. All definitions should be added by `target_compile_definitions(opentelemetry_api INTERFACE ...)` to be expoted by cmake.
We should avoid to use `add_definitions(...)` unless the definitions is only used by unit tests, should I create another issue to fix it? @lalitb @ThomsonTan

**Comment by owent** [comment]
> LGTM in general, @owent may want to have a look into this, specifically the batch span/log processor part. Also, we should add a CI action to test the async feature ( e.g, add another test similar to [bazel_test](https://github.com/open-telemetry/opentelemetry-cpp/blob/a7e814a632fdaee0c76246358ffad811ac06a7e9/.github/workflows/ci.yml#L131) which builds using ENABLE_ASYNC_EXPORT macro).

Agree, I think we can just add `-DENABLE_ASYNC_EXPORT` for bazel options and add `-DENABLE_ASYNC_EXPORT=ON` for cmake options in `do_ci.sh/do_ci.ps1`

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h
Comment (owent) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:65]: Should `if (is_export_async_ == false)` only be checked when `ENABLE_ASYNC_EXPORT` is defined?

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h
Comment (owent) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:37]: `is_export_async` will not be declared without `ENABLE_ASYNC_EXPORT` in `BatchSpanProcessorOptions`. I think we should keep the same behaviour with `SimpleSpanProcessor` and just remove `is_export_async_`.

File: sdk/src/logs/batch_log_processor.cc
Comment (owent) [inline sdk/src/logs/batch_log_processor.cc:26]: `is_export_async` will not be declared without `ENABLE_ASYNC_EXPORT` in `BatchSpanProcessorOptions`. I think we should also remove `is_export_async_` here.

File: sdk/include/opentelemetry/sdk/logs/batch_log_processor.h
Comment (owent) [inline sdk/include/opentelemetry/sdk/logs/batch_log_processor.h:49]: Should we also remove `is_export_async_` in `BatchLogProcessor` and `BatchSpanProcessor` ?

File: sdk/src/logs/simple_log_processor.cc
Comment (owent) [inline sdk/src/logs/simple_log_processor.cc:53]: Just like in `SimpleSpanProcessor`, `if (is_export_async_ == false)` should only be checked when `ENABLE_ASYNC_EXPORT` is defined, or we should always call sync Export.

File: sdk/src/trace/batch_span_processor.cc
Comment (owent) [inline sdk/src/trace/batch_span_processor.cc:221]: Just like in `SimpleSpanProcessor`, `if (is_export_async_ == false)` should only be checked when `ENABLE_ASYNC_EXPORT` is defined, or we should always call sync Export.

File: sdk/src/logs/batch_log_processor.cc
Comment (owent) [inline sdk/src/logs/batch_log_processor.cc:238]: Just like in `SimpleSpanProcessor`, `if (is_export_async_ == false)` should only be checked when `ENABLE_ASYNC_EXPORT` is defined, or we should always call sync Export.

**Review by owent — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (DebajitDas) [reply CMakeLists.txt:231]: Added.

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_http_exporter.h
Comment (DebajitDas) [reply exporters/otlp/include/opentelemetry/exporters/otlp/otlp_http_exporter.h:92]: Done

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_http_log_exporter.h
Comment (DebajitDas) [reply exporters/otlp/include/opentelemetry/exporters/otlp/otlp_http_log_exporter.h:93]: Done

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h
Comment (DebajitDas) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:69]: Done

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h
Comment (DebajitDas) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:35]: Done

File: sdk/src/logs/batch_log_processor.cc
Comment (DebajitDas) [reply sdk/src/logs/batch_log_processor.cc:26]: Done

File: sdk/include/opentelemetry/sdk/logs/batch_log_processor.h
Comment (DebajitDas) [reply sdk/include/opentelemetry/sdk/logs/batch_log_processor.h:49]: Done

File: sdk/src/logs/simple_log_processor.cc
Comment (DebajitDas) [reply sdk/src/logs/simple_log_processor.cc:55]: Done

File: sdk/src/trace/batch_span_processor.cc
Comment (DebajitDas) [reply sdk/src/trace/batch_span_processor.cc:217]: Done

File: sdk/src/logs/batch_log_processor.cc
Comment (DebajitDas) [reply sdk/src/logs/batch_log_processor.cc:235]: Done

**Review by owent — APPROVED** [review]
LGTM.
The rest of the things is adding CI actions to test the async feature.(with and without)

**Comment by DebajitDas** [comment]
Could this PR be merged? I would add max export changes once this is merged.

**Comment by owent** [comment]
@lalitb I think this PR can be merged right now. I'm not in [@open-telemetry/cpp-approvers](https://github.com/orgs/open-telemetry/teams/cpp-approvers) right now, my approval can not let the process continue.

**Comment by lalitb** [comment]
> We should avoid to use add_definitions(...) unless the definitions is only used by unit tests, should I create another issue to fix it? @lalitb @ThomsonTan

Good point. Yes please create a issue to fix it @owent

**Comment by lalitb** [comment]
> @lalitb I think this PR can be merged right now. I'm not in [@open-telemetry/cpp-approvers](https://github.com/orgs/open-telemetry/teams/cpp-approvers) right now, my approval can not let the process continue.

Thanks, Have added you to approvers now. Seems got missed somehow.

**Comment by lalitb** [comment]
@DebajitDas  - Merging this, the CI test and max_export can be added in separate PR.

**Pull request merged**
